### PR TITLE
C#: Extract function pointers

### PR DESCRIPTION
**Pull request by tamasvajk**
This PR adds support for function pointer type extraction from source code. Calls to function pointers are also extracted, but only a thin `FunctionPointerCall` class is added for the time being.

Missing:
- [x] DB upgrade folder
- [x] Stats job https://jenkins.internal.semmle.com/job/Stats/job/CSharp/30/
- [x] Differences job https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/788/
- [x] update release notes date
- [x] Investigate if fnptr conversions are correctly extracted: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/function-pointers#function-pointer-conversions

### CODE REVIEW

File: csharp/ql/test/library-tests/csharp9/FunctionPointer.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            return 0;
+        }
+
+        static void M1(delegate*<ref int, out object?, int> f)
+        {
+            int i = 42;
+            int j = f(ref i, out object? o);
+        }
+
+        static void M2<T>(delegate* unmanaged[Stdcall/*, StdcallSuppressGCTransition*/]<ref int, out object?, T, void> f) where T : new()
```
Comment (tamasvajk) [inline csharp/ql/test/library-tests/csharp9/FunctionPointer.cs:22]: I think we need dotnet 5 for `StdcallSuppressGCTransition`.

File: csharp/ql/src/semmle/code/csharp/exprs/Expr.qll [hunk 2]
```diff
@@ -562,7 +562,7 @@ class Cast extends Expr {
   TypeAccess getTypeAccess() { result = this.getChild(1) }
 
   /** Gets the type that the underlying expression is being cast to. */
-  Type getTargetType() { result = this.getTypeAccess().getTarget() }
+  Type getTargetType() { result = this.getType() }
```
Comment (tamasvajk) [inline csharp/ql/src/semmle/code/csharp/exprs/Expr.qll:565]: This is a slightly unrelated change. Implicit casts have no `type access`, and therefore they had no target type either. I think when we have a `type access`, it will always match the expression type, so for normal casts this change has no impact. For implicit casts the change would return the type to which we convert. Let me know if my assumption here is incorrect.

**Comment by tamasvajk** [comment]
Fails with BDD node limit reached error.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Types/FunctionPointerType.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+        public override void Populate(TextWriter trapFile)
+        {
+            var unmanagedCallingConventionTypes = symbol.Signature.UnmanagedCallingConventionTypes.Select(nt => Create(Context, nt)).ToArray();
+
+            trapFile.function_pointer_calling_conventions(this, (int)symbol.Signature.CallingConvention);
+            for (var i = 0; i < unmanagedCallingConventionTypes.Length; i++)
+            {
+                trapFile.has_unmanaged_calling_conventions(this, i, unmanagedCallingConventionTypes[i].TypeRef);
+            }
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Types/FunctionPointerType.cs:30]:
```suggestion
            trapFile.function_pointer_calling_conventions(this, (int)symbol.Signature.CallingConvention);

            foreach (var (conv, i) in symbol.Signature.UnmanagedCallingConventionTypes.Select((nt, i) => (Create(Context, nt), i)))
            {
                trapFile.has_unmanaged_calling_conventions(this, i, conv.TypeRef);
            }
```
this lets you get rid of the intermediate array construction.

File: csharp/ql/src/semmle/code/csharp/Type.qll [hunk 4]
```diff
@@ -797,6 +797,31 @@ class DelegateType extends RefType, Parameterizable, @delegate_type {
...
+ * ```csharp
+ * delegate*<int, void>
+ * ```
+ */
+class FunctionPointerType extends Type, Parameterizable, @function_pointer_type {
+  /** Gets the return type of this function pointer. */
+  Type getReturnType() { function_pointer_return_type(this, getTypeRef(result)) }
+
+  /** Gets the calling convention. */
+  int getCallingConvention() { function_pointer_calling_conventions(this, result) }
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/Type.qll:812]: We should wrap the `int` in a class layer that covers the various calling conventions (unless there are just too many?)

**Review by hvitved — CHANGES_REQUESTED** [review]
Looks great to me. Some minor comments.

File: csharp/ql/src/semmle/code/csharp/Type.qll (hunk 4 above)
Comment (tamasvajk) [reply csharp/ql/src/semmle/code/csharp/Type.qll:812]: I think there are 6 options. What do you mean by wrapping the `int`? Do you mean having 6 dedicated predicates, like in `TypeParameterConstraints` for `general_type_parameter_constraints()`?

Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/Type.qll:812]: No, what I meant was something like
```ql
private newtype TCallingConvention = MkCallingConvention(int i) { function_pointer_calling_conventions(_, i) }

class CallingConvention extends TCallingConvention { }

// Add sub classes

CallingConvention getCallingConvention() { exists(int i | function_pointer_calling_conventions(this, i) and result = MkCallingConvention(i) }
```

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Invocation.cs [hunk 5]
```diff
@@ -153,9 +164,26 @@ private static bool IsDelegateCall(ExpressionNodeInfo info)
                 return true;
             }
 
-            return si.Symbol != null &&
-                si.Symbol.Kind == SymbolKind.Method &&
-                ((IMethodSymbol)si.Symbol).MethodKind == MethodKind.DelegateInvoke;
+            return check(si.Symbol);
+        }
+
+        private static bool IsDelegateLikeCall(ISymbol symbol)
```
[the remark this answers is not in the document]
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Invocation.cs:170]: We have this nested ternary operator a couple of lines below:
```
private static ExprKind GetKind(ExpressionNodeInfo info)
{
    return IsNameof((InvocationExpressionSyntax)info.Node)
        ? ExprKind.NAMEOF
        : IsDelegateLikeCall(info)
            ? IsDelegateInvokeCall(info)
                ? ExprKind.DELEGATE_INVOCATION
                : ExprKind.FUNCTION_POINTER_INVOCATION
            : IsLocalFunctionInvocation(info)
                ? ExprKind.LOCAL_FUNCTION_INVOCATION
                : ExprKind.METHOD_INVOCATION;
}
```
And I think having dedicated methods for `IsDelegateLikeCall` and `IsDelegateInvokeCall` is quite useful in making this readable.

[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Invocation.cs:170]: But that calls `IsDelegateLikeCall(ExpressionNodeInfo info)` and not `IsDelegateLikeCall(ISymbol symbol)`, right? I was only suggesting inlining the latter.

**Comment by tamasvajk** [comment]
@hvitved I applied your findings and also pushed an extra commit with changes to query the annotated return type of function pointers.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Invocation.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Invocation.cs]: Ahh, okay, I understand now.

File: csharp/ql/src/semmle/code/csharp/Type.qll [hunk 6]
```diff
@@ -797,6 +797,63 @@ class DelegateType extends RefType, Parameterizable, @delegate_type {
...
+}
+
+/** Managed calling convention for passing extra arguments. */
+class VarArgsCallingConvention extends CallingConvention {
+  VarArgsCallingConvention() { this = MkCallingConvention(5) }
+
+  override string toString() { result = "VarArgsCallingConvention" }
+}
+
+// Add sub classes
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/Type.qll:857]: Remove :-)

File: csharp/ql/src/semmle/code/csharp/Type.qll [hunk 7]
```diff
@@ -797,6 +797,63 @@ class DelegateType extends RefType, Parameterizable, @delegate_type {
...
 
+private newtype TCallingConvention =
+  MkCallingConvention(int i) { function_pointer_calling_conventions(_, i) }
+
+/**
+ * Represents a signature calling convention. Specifies how arguments in a given
+ * signature are passed from the caller to the callee.
+ */
+class CallingConvention extends TCallingConvention {
+  string toString() { result = "CallingConvention" }
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/Type.qll:809]: Add `/** Gets a textual representation of this calling convention. */`

**Review by hvitved — CHANGES_REQUESTED** [review]

**Review by hvitved — APPROVED** [review]
